File: integrations/csharp/sentry/SentrySink.cs
using Axora.Logger.Sinks;

namespace Axora.Logger.Sentry;

public interface ISentryScopeBridge
{
    void SetTag(string key, string value);
    void SetExtras(IDictionary<string, object?> extras);
    void SetLevel(string level);
}

public interface ISentryBridge
{
    void WithScope(Action<ISentryScopeBridge> callback);
    void CaptureException(Exception exception);
    void AddBreadcrumb(string category, string message, string level, DateTimeOffset timestamp, IDictionary<string, object?> data);
}

public sealed class SentrySink : ISink
{
    private readonly ISentryBridge _bridge;

    public SentrySink(ISentryBridge bridge)
    {
        _bridge = bridge;
    }

    public Task ExportAsync(WideEventPayload @event)
    {
        if (@event.Status is "error" or "timeout")
        {
            _bridge.WithScope(scope =>
            {
                scope.SetTag("service", @event.Service);
                scope.SetTag("environment", @event.Environment);
                scope.SetTag("operation", @event.Operation);
                scope.SetTag("axora.event_id", @event.EventId);
                scope.SetExtras(@event.Context);
                scope.SetLevel(@event.Level == "fatal" ? "fatal" : @event.Level);
            });

            var error = new InvalidOperationException(@event.Error.Message ?? @event.Operation);
            _bridge.CaptureException(error);
            return Task.CompletedTask;
        }

        var data = new Dictionary<string, object?>(@event.Context)
        {
            ["duration_ms"] = @event.DurationMs,
            ["axora_event_id"] = @event.EventId,
        };

        _bridge.AddBreadcrumb(
            @event.Operation,
            @event.Operation,
            @event.Level,
            DateTimeOffset.Parse(@event.TimestampStart),
            data);

        return Task.CompletedTask;
    }
}

File: integrations/typescript/sentry/src/SentrySink.ts
import type { Sink, WideEventPayload } from '@axora/logger-core';
import * as Sentry from '@sentry/node';

function severityLevel(level: WideEventPayload['level']): Sentry.SeverityLevel {
  return level as Sentry.SeverityLevel;
}

export class SentrySink implements Sink {
  async export(event: WideEventPayload): Promise<void> {
    const isError = event.status === 'error' || event.status === 'timeout';

    if (isError) {
      Sentry.withScope((scope) => {
        scope.setTag('service', event.service);
        scope.setTag('environment', event.environment);
        scope.setTag('operation', event.operation);
        scope.setTag('axora.event_id', event.event_id);
        scope.setExtras(event.context);
        scope.setLevel(event.level === 'fatal' ? 'fatal' : severityLevel(event.level));

        const error = new Error(event.error.message ?? event.operation);
        error.name = event.error.type ?? 'AxoraError';
        if (event.error.stack) {
          error.stack = event.error.stack;
        }

        Sentry.captureException(error);
      });
      return;
    }

    Sentry.addBreadcrumb({
      category: event.operation,
      message: event.operation,
      level: severityLevel(event.level),
      data: {
        ...event.context,
        duration_ms: event.duration_ms,
        axora_event_id: event.event_id,
      },
      timestamp: Date.parse(event.timestamp_start) / 1000,
    });
  }
}
